### PR TITLE
Extended package to be a multimodal benchmark

### PR DESCRIPTION
**Pull request by x-tabdeveloping**
Multiple contributions:
1. Refactored code base:
    - Convenience functions
    - Better data types and schemata
    - Better naming
    - `msgspec` for serialization as it is way faster and validates data
2. Added a bunch of multimodal datasets from MBEIR
3. Added MTEB API implementations of multimodal encoders, which makes them easy to switch out.
4. Added multimodal models, implemented a double-sided $S^3$ so that negative topics also get evaluated
5. Added implementation of MultimodalZeroShotTM and MultimodalContrastTM from [this paper](https://arxiv.org/pdf/2403.17308)
6. Implemented two naturally multimodal metrics, which are hard to balance and are much more representative of modern topic models:
    - Embedding Silhouette score on topic representations (words, documents and optionally images)
    - Corpus embedding coverage metric
7. Added an integration test suite to ensure that the benchmark doesn't break with new additions

### CODE REVIEW

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    steps:
+      - uses: actions/checkout@v4
+      - name: Set up Python ${{ matrix.python-version }}
+        uses: actions/setup-python@v4
+        with:
+          python-version: ${{ matrix.python-version }}
+          cache: "pip"
+      # You can test your matrix by printing the current Python version
+      - name: Display Python version
+        run: python3 -c "import sys; print(sys.version)"
```
Comment (KennethEnevoldsen) [inline .github/workflows/tests.yml:29]:
```suggestion
```

File: .github/workflows/tests.yml [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+      - name: Install dependencies
+        run: pip install . && pip install pytest
+      - name: Run tests
+        run: python3 -m pytest tests/
```
Comment (KennethEnevoldsen) [inline .github/workflows/tests.yml:34]: This could break with changes to pytest, as you don't version it (you can use an optional dependency or or the ideal is probably to use [dependency group](https://peps.python.org/pep-0735/))

File: scripts/multimodal/run_multimodal.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
+import os
+from pathlib import Path
+
+os.chdir(Path(__file__).parent.parent)
```
Comment (KennethEnevoldsen) [inline scripts/multimodal/run_multimodal.py:4]: would probably not change the directory, but probably just do:

```suggestion
repo_path = Path(__file__).parent.parent
```

and use that to construct your out_folder

File: scripts/multimodal/run_multimodal.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+from sklearn.feature_extraction.text import CountVectorizer
```
Comment (KennethEnevoldsen) [inline scripts/multimodal/run_multimodal.py:6]: some unused imports

File: tests/test_integration.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+
+@dataset_registry.register("dummy_20NG")
+def load_newsgroups_raw() -> list[str]:
+    ds = fetch_20newsgroups(subset="all")
+    corpus = list(ds.data)[:32]
+    return Dataset(corpus)
+
+
+OUT_FOLDER = Path(__file__).parent.joinpath("__test_results")
```
Comment (KennethEnevoldsen) [inline tests/test_integration.py:35]:
```suggestion
OUT_FOLDER = Path(__file__).parent / "__test_results"
```
Not sure if you know that this is an option?

File: tests/test_integration.py [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+METRICS = ["wec_in", "coverage", "diversity", "silhouette", "stop_freq"]
```
Comment (KennethEnevoldsen) [inline tests/test_integration.py:37]: move constants to the top?

File: topic_benchmark/metrics/multimodal.py [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+        pass
+    embeddings = np.array(embeddings)
+    labels = np.array(labels)
+    return embeddings, labels
+
+
+@metric_registry.register("silhouette")
+def load_silhouette() -> Metric:
+    top_k = 10
+    encoder = CLIPModelWrapper("openai/clip-vit-base-patch32")
```
Comment (KennethEnevoldsen) [inline topic_benchmark/metrics/multimodal.py]: would move to default arguments (seems like stuff that you might want to change)

File: topic_benchmark/metrics/multimodal.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+
+    def score(data: TopicData, dataset_name: Optional[str]):
+        embeddings, labels = encode_descriptives(data, top_k, encoder)
+        res = silhouette_score(embeddings, labels, metric="cosine")
```
Comment (KennethEnevoldsen) [inline topic_benchmark/metrics/multimodal.py:55]: no idea if this is a good score - probably not here that we should discuss it, but I would love to be convinced on the metric side (might be worth taking some time to do that)

Comment (KennethEnevoldsen) [reply topic_benchmark/metrics/multimodal.py:55]: Let us do that for all the metrics - do they use metrics in the two multimodal TM papers that we want to baseline against?

**Review by KennethEnevoldsen — APPROVED** [review]
Generally looks good. I have only made minor suggestions

1) Haven't reviewed models implementation, but selection looks good. We should probably write up a rationale for choosing these ones specifically (encoders as well)
2) For the datasets I imagine that there is a rationale for the selection. Would love to hear it. It looks decent though no immediate concerns
3) Let us do a discussion on metrics (see below)

File: scripts/multimodal/run_multimodal.py (hunk 3 above)
Comment (x-tabdeveloping) [reply scripts/multimodal/run_multimodal.py:4]: There is a reason for doing this. I want the script to import code from the repo not from the package, and if you are in the root directory of the repo running:
```bash
python3 scripts/multimodal/run_benchmark.py
```
it will fail to import stuff

File: tests/test_integration.py (hunk 5 above)
Comment (x-tabdeveloping) [reply tests/test_integration.py:35]: Yes I have known about it for a while. I just don't like it for some reason. Like in my head it doesn't make sense that the same operator is used for division as for adding paths together, like shouldn't this be a `+` instead?? 

Anyway... I prefer to spell it out, it doesn't hurt anyone

File: .github/workflows/tests.yml (hunk 2 above)
Comment (x-tabdeveloping) [reply .github/workflows/tests.yml:34]: It could, but it probably won't. And when it does, it will be two minutes to fix it. I think it's ok

File: topic_benchmark/metrics/multimodal.py (hunk 7 above)
Comment (x-tabdeveloping) [reply topic_benchmark/metrics/multimodal.py]: DUN

File: scripts/multimodal/run_multimodal.py (hunk 3 above)
Comment (KennethEnevoldsen) [reply scripts/multimodal/run_multimodal.py:4]: But shouldn't it be the same? Otherwise I think that is a problem

Also if that is the case `sys.path.append` is your way to do that without changing your directory
